[IV-23-24] Objetivo 5: Creación de un contenedor para pruebas

johnwaves opened the pull request:
* [x] ¿Se han establecido criterios previos de elección de la imagen base, así
como los criterios de búsqueda que han conducido a las imágenes que se están
considerando?
* [x] ¿Se ha tenido en cuenta que la imagen es para testear, y que por tanto
tiene una función específica, tal como testear cómo va nuestra aplicación en la
última versión publicada del lenguage?
* [x] ¿Se ha tenido en cuenta que la versión de la imagen base sea la última?
* [x] ¿Se han tenido en cuenta criterios reales para la elección, y no criterios
irrelevantes como cuanto tiempo se tarda en construirla? (Se va a construir una
vez, se va a usar cientos o miles de veces)
* [x] ¿Se han seguido las buenas prácticas en las órdenes que construyen
   la imagen? ¿Especialmente las relativas a usuarios no privilegiados y demás?
* [x] ¿La imagen incluye sólo y exclusivamente la infraestructura
   necesaria para pasar los tests? Por ejemplo, si se ha elegido la imagen
   "oficial" (lo que hay que hacer sólo en casos muy justificados) ¿se ha
   comprobado que lo que se instale sea sólo y exclusivamente lo necesario para
   nuestro proyecto?
* [x] ¿Se han documentado y enlazado los commits a las imágenes que se
   han testeado?
* [x] ¿Está establecido correctamente como `WORKDIR` el que se va a usar en los
   tests que se van a lanzar?
* [x] ¿Se usa algún directorio adecuado para copiar contenidos del repositorio
en vez del directorio de trabajo?
* [x] ¿Se han establecido claramente los criterios de búsqueda y de aceptación de
   la imagen base?
* [x] ¿Tienes claro cuales son las condiciones en las que tiene que actualizarse
la imagen en DockerHub y has configurado el workflow de acuerdo con ello? En
particular, ¿sabes qué condiciones tienen que darse para que se actualice y qué
condiciones *no* deben darse?
* [x] Si se usan GitHub actions para su publicación, ¿se han considerado
cuidadosamene los criterios (ramas, paths) para que se lancen?
* [x] ¿Se ha comprobado que el ENTRYPOINT del contenedor no escriba nada en el
directorio que monta, ya que no tiene derecho de escritura?
* [x] ¿Se ha abierto un shell en el contenedor una vez construida la imagen y
comprobado que los ficheros que se necesitan están todos incluidos, y que no hay
ficheros que no se necesiten?

johnwaves committed:
Actualiza README

johnwaves committed:
Merge branch 'main' into Objetivo-5

johnwaves committed:
Añade criterios de selección #55

johnwaves committed:
Actualiza versión Go closes #56

La versión de Go se actualiza a la última versión
soportada por la imagen seleccionada para las pruebas.

johnwaves committed:
Configura Dockerfile #55

johnwaves committed:
Actualiza información principal closes #55

johnwaves committed:
Corrige ubicación workflows #55

johnwaves committed:
Añade enlace para criterios de elección

johnwaves committed:
Corrige sintaxis código

johnwaves committed:
Corrige nuevamente código

JJ reviewed (approved):
:+1:

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+FROM bitnami/golang:latest
+
+RUN adduser --disabled-password -u 1001 test
```
¿Por qué deshabilitas la clave? ¿O le asignas el usuario 1001 a test? El que se use el usuario 1001 es sólo un ejemplo, diferentes sistemas de CI pueden tener diferentes UIDs.

johnwaves replied:
Deshabilité la clave porque tenía entendido que al crear usuarioos que fuesen temporales para pruebas o desarrollo no era necesario que tengan una contraseña.

JJ replied:
¿En qué issue está eso incluido? Si le dejas la contraseña, ¿en qué cambiaría la operación de este imagen para ejecutar las pruebas?

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+
+WORKDIR /app
+
+ENV GOPROXY=https://proxy.golang.org,direct
```
¿Esto por qué lo haces?

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+WORKDIR /app
+
+ENV GOPROXY=https://proxy.golang.org,direct
+ENV GOSUMDB=sum.golang.org
```
Como el resto del fichero, lo has [creado aquí](https://github.com/johnwaves/recambios-express/commit/52eed479dda882302ecc38fb3f9279dbacc72fe7) enlazando el issue #55, que no comenta qué problemas hay que resolver mediante esta variable de entorno.

JJ commented on go.mod:
```diff
@@ -1,11 +1,12 @@
 require (
 	github.com/davecgh/go-spew v1.1.1 // indirect
 	github.com/pmezard/go-difflib v1.0.0 // indirect
+	github.com/stretchr/objx v0.5.1 // indirect
```
Si no hay ningún módulo nuevo en este objetivo, ¿por qué se añade esto?

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+ENV GOPROXY=https://proxy.golang.org,direct
+ENV GOSUMDB=sum.golang.org
+
+RUN go install github.com/go-task/task/v3/cmd/task@latest
```
¿Qué ventaja tiene instalarlo así frente a usar el binario?

johnwaves replied:
Tal vez por el hecho der ser más directo y sencillo.

JJ replied:
¿En qué sentido es esto más directo que descargarse el binario y colocarlo en un sitio donde se pueda ejecutar?

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+
+RUN go install github.com/go-task/task/v3/cmd/task@latest
+
+COPY go.mod go.sum Taskfile.yml ./
```
¿Por qué necesitas el Taskfile aquí?

johnwaves replied:
Realmente no es necesario para esa orden, lo cambio.

JJ reviewed (changes requested):
En un segundo vistazo, quizás deberías darle una pequeña vuelta. Y también me sentiría más cómodo si no usaras la imagen de bitnami elegida. Hay muchas otras imágenes que se pueden usar justificadamente.

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+FROM bitnami/golang:latest
```
Entonces, según #56, cada vez que cambie la versión en latest tendrás que cambiar go.mod, ¿no?
*Eso* por definición es deuda técnica.

johnwaves committed:
Modifica imagen elegida closes #57

Se usa otra imagen diferente para el contenedor de
pruebas. Se actualizan el documento con los criterios.

johnwaves committed:
Reelige imagen closes #57

La imagen que se elige esta vez solventa los problemas
ocurridos usando BusyBox, como la falta de capacidad
por cumplir con los requisitos del objetivo.

johnwaves committed:
Configura Dockerfile closes #58

La configuración del Dockerfile se basa en dos partes:
la primera es la construcción del entorno con Go
para poder compilar el código fuente, y la segunda
es la configuración del entorno de ejecución (Debian).

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,25 @@
+
+COPY --from=build /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/
+
+RUN adduser -u 1001 test
```
¿por qué usas el UID 1001?

johnwaves replied:
Si le asigno el UID 1001 al usuario test, me aseguro que el contenedor se ejecuta con ese usuario y su entorno preconfigurado, con los permisos necesarios para utilizar una caché para almacenar los resultados de las pruebas. Si no lo hago así, cuando ejecuto ```docker run -u 1001 -t -v `pwd`:/app/test/ johnwaves1/recambios-express``` me muestra que el acceso está denegado.

JJ replied:
Precisamente por eso es por lo que tienes que usar el UID por omisión. Si el acceso está denegado, es que no estás cubriendo el objetivo.

JJ commented on README.md:
```diff
@@ -26,8 +26,25 @@ Para automatizar las tareas se utiliza **Task**. [Aquí](https://github.com/john
+## Contenedor para pruebas
+
+### Imagen para el contenedor de pruebas
+**Bitnami** es la imagen seleccionada para la ejecución de los tests unitarios sobre este proyecto. Tanto los criterios seguidos como las demás imágenes candidatas se detallan [aquí](https://github.com/johnwaves/recambios-express/blob/Objetivo-5/docs/eleccion_images.md).
```
Esta no es la imagen seleccionada, ¿no?

JJ commented on go.mod:
```diff
@@ -1,6 +1,6 @@
 module recambios-express
 
-go 1.18
+go 1.21.7
```
Sigo pensando que cambiar la versión por el contenedo no es una buena idea. [¿No hay imágenes con Golang 1.18?](https://hub.docker.com/_/golang/tags?page=1&name=1.18)

johnwaves replied:
Sí que las hay, pero el grado de [vulnerabilidad](https://hub.docker.com/layers/library/golang/1.18-bullseye/images/sha256-740324e52de766f230ad7113fac9028399d6e03af34883de625dc2230ef7927e?context=explore) que presentan es mayor en comparación con versiones más recientes,

JJ replied:
No me hagas trampas en el solitario. Ninguna de esas vulnerabilidades se debe a la versión de go. Mira tú mismo las capas que introducen vulnerabilidades. Si estás construyendo en dos fases, ¿cuales de esas vulnerabilidades te afectan?

JJ commented on docs/eleccion_images.md:
```diff
@@ -0,0 +1,32 @@
+### [Golang con Alpine](https://hub.docker.com/_/alpine)
+Esta es una de las opciones que ofrece un entorno minimalista y eficiente para desplegar servicios ligeros escritos en Go, dado que incorpora un compilador y algunas herramientas para desarrollar y ejecutar aplicaciones escritas en este lenguaje (como pueden ser `Go fmt`, `Go test` o `Go build`). Pese a disponer de estas ventajas, cabe mencionar que Alpine tiene un repositorio aún creciente y el  ...
+
+### [Bitnami](https://hub.docker.com/r/bitnami/golang)
```
Estás mezclando imágenes que tienen solo el sistema operativo con imágenes con Go instalado. Y no estás mencionando, por ejemplo, la imagen "oficial" que es la que estás usando.

JJ reviewed (changes requested):
Está bien que uses una construcción en dos fases, pero en ningún momento lo has justificado en la elección de las imágenes base. 
Aparte, ¿`go.sum` y `go.mod` tienen que quedarse en la imagen?
Aparte, ¿has mirado las vulnerabilidades que tiene la imagen que estás usando? ¿Tienen alguna importancia de por sí o en la aplicación en la que la estás usando?
Por otro lado, te agradecería que tuvieras esto listo antes del final de la mañana. La presentación tendría que ser mañana por la mañana (por la tarde no estoy disponible), así que convendría que lo supiéramos con cierto tiempo de antelación.

johnwaves commented:
> Está bien que uses una construcción en dos fases, pero en ningún momento lo has justificado en la elección de las imágenes base. Aparte, ¿`go.sum` y `go.mod` tienen que quedarse en la imagen? Aparte, ¿has mirado las vulnerabilidades que tiene la imagen que estás usando? ¿Tienen alguna importancia de por sí o en la aplicación en la que la estás usando? Por otro lado, te agradecería que tuvieras esto listo antes del final de la mañana. La presentación tendría que ser mañana por la mañana (por la tarde no estoy disponible), así que convendría que lo supiéramos con cierto tiempo de antelación.

Voy a corregirlo y añadir lo que me faltó por explicar de la imagen. No he podido verlo antes por el trabajo.

johnwaves commented:
> Está bien que uses una construcción en dos fases, pero en ningún momento lo has justificado en la elección de las imágenes base. Aparte, ¿`go.sum` y `go.mod` tienen que quedarse en la imagen? Aparte, ¿has mirado las vulnerabilidades que tiene la imagen que estás usando? ¿Tienen alguna importancia de por sí o en la aplicación en la que la estás usando? Por otro lado, te agradecería que tuvieras esto listo antes del final de la mañana. La presentación tendría que ser mañana por la mañana (por la tarde no estoy disponible), así que convendría que lo supiéramos con cierto tiempo de antelación.

Comprobándolo nuevamente, tal vez podría prescindir de incluir `go.mod` en la imagen dado que el `ENTRYPOINT` es `["task", "test"]` y Task ya maneja la descarga y gestión de las dependencias. Por otra parte, `go.sum` también podría no incluirse, aunque por seguridad he leído que resulta una buena práctica dejarlo.

johnwaves committed:
Corrige nombre imagen #57

Se corrige el nombre de la imagen utilizada y
modifica la orden para la ejecución del módulo de
test.

johnwaves committed:
Modifica Dockerfile #58

Se prescinde de copiar go.mod y go.sum en el contenedor,
ya que no son necesarios para la ejecución de
la aplicación. Asimismo se elimina la orden go mod
download, pues mediante Task descargan todas las
dependencias necesarias.

johnwaves committed:
Modifica la elección de la imagen #57

Se especifican las imágenes que vienen con Go instalado
y se detalla por qué se usa la elegida y su proceso
de construcción.

JJ reviewed (changes requested):
Lo siento, pero de lo que se trata es de que se pueda ejecutar el test con un usuario cualquiera. Sustituye 1001 en la ejecución del contenedor por 1111. Si no funciona, es que no lo estás haciendo correctamente. Hacer el truco de crear un usuario con el mismo UID no cubre los objetivos, porque no has diagnosticado correctamente el problema, ni lo has solucionado. Tampoco deberías haber cerrado el comentario sin reaccionar.

johnwaves committed:
Modifica configuración existente closes #59

En la solución que se implementa se crea el
directorio /cache con permisos de escritura para que
cualquier usuario, independientemente de su UID, pueda
escribir. A continuación, mediante las variables de
entorno GOCACHE y GOPATH se redirecciona la salida al
directorio /cache para asegurar que se tiene un lugar
para almacenar y acceder a los datos temporales sin
la necesidad de disponer de permisos de escritura en
el directorio principal.

johnwaves commented:
> Lo siento, pero de lo que se trata es de que se pueda ejecutar el test con un usuario cualquiera. Sustituye 1001 en la ejecución del contenedor por 1111. Si no funciona, es que no lo estás haciendo correctamente. Hacer el truco de crear un usuario con el mismo UID no cubre los objetivos, porque no has diagnosticado correctamente el problema, ni lo has solucionado. Tampoco deberías haber cerrado el comentario sin reaccionar.

Solucionado. La ejecución se puede llevar a cabo usando cualquier UID.

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,24 @@
+
+RUN mkdir /app /cache && chmod -R 777 /cache
+ENV GOCACHE=/cache
+ENV GOPATH=/cache
```
¿Seguro que es una buena idea hacer esto? ¿Lo que instales de Go se va a instalar aquí en este directorio? ¿GOPATH no debería estar también en el path?

johnwaves replied:
Así es, GOPATH debería estar en el path. Y lo que se instale lo hará en /cache, por tanto, he de cambiarlo.

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,24 @@
+ENV GOCACHE=/cache
+ENV GOPATH=/cache
+
+RUN adduser test
```
Habiendo un usuario, ¿cuál es la práctica habitual para colocar `GOPATH`?

johnwaves replied:
La práctica habitual es colocar el `GOPATH` en el directorio del usuario. Lo modifico y lo actualizo en breve.

JJ replied:
Vale que esto esté fuera de la línea modificada, pero debes resolver todos los comentarios. Y ahora perdona que me ponga mijicas, pero estás creando el usuario *después* del home (dos líneas más arriba).

JJ reviewed (changes requested):
Evidentemente, has solucionado los problemas que habías encontrado, pero deberías intentar no incluir otros.

JJ commented:
Y te rogaría que solucionaras esto esta misma mañana, antes de las 2 de la tarde. Ya estamos en plazo extra, porque hiciste la entrega el domingo (el último día era el miércoles pasado) y ya se ha pasado la fecha de evaluación, que era ayer.

JJ commented:
Aparte, como ves, no está funcionando el workflow para construir la imagen. De hecho, ahora mismo me  estoy dando cuenta de que no ha funcionado nunca porque no lo tienes bien configurado.

johnwaves committed:
Actualiza configuración de Dockerfile #58

En esta modificación se coloca GOPATH en el directorio
del usuario, así se tendrán los permisos necesarios
para leer y escribir en el mismo. Además,
también se añade GOPATH/bin al PATH para que
se pueda ejecutar cualquier binario instalado.

johnwaves committed:
Actualiza configuración del workflow #60

Mediante esta modificación del archivo, se permite
el funcionamiento del workflow de Docker. La orden
build no es necesaria.

johnwaves committed:
Cambia de nuevo configuración closes #60

Se cambia el campo del TOKEN por la contraseña para
posibilitar el acceso a Docker Hub.

johnwaves committed:
Modifica versión finial closes #58

johnwaves commented:
@JJ, la GitHub action ya pasa los checks. El problema ocurría al utilizar `TOKEN` en vez de `PASSWORD`. Listo para revision.

JJ commented on Dockerfile:
```diff
@@ -0,0 +1,27 @@
+ENV GOCACHE=/home/test/.cache/go-build
+ENV PATH="$GOPATH/bin:$PATH"
+
+RUN mkdir -p "$GOPATH" "$GOCACHE" && chmod -R 777 "$GOPATH" "$GOCACHE"
```
Si `GOPATH` tiene estos permisos, ¿cualquier persona que monte esto puede borrar sus contenidos o añadir otros? Es una buena práctica añadir sólo los permisos que sean estrictamente necesarios. También estás creando estos path antes de crear el usuario, así que el propietario seguirá siendo ROOT.

johnwaves replied:
He estado intentando arreglar esto desde que escribió el comentaro pero no doy pie con bola, no sé exactamente cómo configurarlo dentro del directorio home y que tenga los permisos adecuados.

JJ reviewed (changes requested):
Créeme que lo siento, pero tendrás que corregir alguna cosa todavía.

johnwaves committed:
Modifica permisos del Dockerfile #58

Los directorios se crean con los permisos 1777,
lo que significa que que restringe la
eliminación/modificación de archivos por parte de otros usuarios
(diferentes al propietario y al root).

johnwaves commented:
Creo que he podido resolverlo @JJ

JJ reviewed (approved):
:+1:
